// vue-modal-service/src/services/ModalService.js
const service = {
    install(Vue) {
        this.event = new Vue();
        const modalService = {
            $active: null,
            $availableModals: [],
            $payload: null,
            $event: this.event,
            $queue: [],
            show: (modalName, payload=null, save=true) => {
                if (modalName === modalService.$active) return;
                if (save && modalService.$active) {
                    modalService.$queue.push({name: modalService.$active, payload: {...modalService.$payload}});
                    modalService.hide(modalService.$active, false);
                }
                modalService.$payload = payload;
                service.event.$emit("show", modalName, payload);
                modalService.$active = modalName;
            },
            hide: (modalName, showLast = true) => {
                service.event.$emit("hide", modalName);
                modalService.$active = null;
                modalService.$payload = null;
                if (showLast && modalService.$queue.length) {
                    const modal = modalService.$queue.pop();
                    modalService.show(modal.name, modal.payload);
                }
            },
            haveInQueue: (modalName) => {
                const index = modalService.$queue.findIndex(({name}) => name === modalName);
                return Number.isInteger(index) && index>=0;
            },
            _registration: (modalName) => {
                if (!modalService.$availableModals.includes(modalName))
                    modalService.$availableModals.push(modalName);
            },
            _unregistration: (modalName) => {
                if (modalService.$availableModals.includes(modalName))
                    modalService.$availableModals.splice(modalService.$availableModals.indexOf(modalName, 1));
            },
        }
        Vue.prototype.$modal = modalService;
    }
}

export default service;
